### PR TITLE
archi/core: correct shared memory and add dynamic shm

### PR DESCRIPTION
**Pull request by etienne-lms**
```
Remove deprecated references to Linux driver resources. This
prevents dependency on Linux OP-TEE driver evolution and opens
perspective to alternate non-secure clients as U-Boot.

Fix Joakim TODO tag on deprecated bootcfg_get_memory().

Add description of the dynamic shared memory: registration/discovery
from secure world and buffer registration by non-secure world.
```

While exploring the new docs, found some untrue statement and deprecated info (incl @jbech-linaro todo tag). This is an attempt to clean the Shared Memory description part while not adding to much info that could deprecate.

### CODE REVIEW

File: architecture/core.rst [hunk 1]
```diff
@@ -800,42 +800,52 @@ Shared Memory
...
+Shared Memory References
 ========================
-The shared memory is allocated by the Linux driver from a pool ``struct
-shm_pool``, the pool contains:
 
-    - The physical address of the start of the pool
-    - The size of the pool
-    - Whether or not the memory is cached
-    - List of chunk of memory allocated.
+The shared memory is allocated and managed by the non-secure world, i.e the
```
Comment (jbech-linaro) [inline architecture/core.rst]: s/i.e/i.e.,/

File: architecture/core.rst [hunk 2]
```diff
@@ -800,42 +800,52 @@ Shared Memory
...
+Linux OP-TEE driver. Secure world only considers the individual shared buffers,
+not their pool. Each shared memory is referenced with associated attributes:
+    - Buffer start address and byte size
+    - Cacheability of the buffer
```
Comment (jbech-linaro) [inline architecture/core.rst]: https://english.stackexchange.com/questions/251159/is-cacheability-a-word-technical-word

File: architecture/core.rst [hunk 3]
```diff
@@ -800,42 +800,52 @@ Shared Memory
...
+    - List of chunks if mapped from discontiguos pages.
```
Comment (jbech-linaro) [inline architecture/core.rst]: s/discontigous/noncontiguous/

File: architecture/core.rst [hunk 4]
```diff
@@ -800,42 +800,52 @@ Shared Memory
...
 
-.. note::
-    - The shared memory pool is physically contiguous.
-    - The shared memory area is **not secure** as it is used by both non-secure
-      and secure world.
+Shared memory buffer references manipulated must fit inside one of the
+shared memory areas known from the OP-TEE core. OP-TEE supports 2 kind
```
Comment (jbech-linaro) [inline architecture/core.rst]: s/2/two/
s/kind/kinds/

File: architecture/core.rst [hunk 5]
```diff
@@ -800,42 +800,52 @@ Shared Memory
...
+of shared memory areas: a mandatory one for physically contiguous buffers
```
Comment (jbech-linaro) [inline architecture/core.rst]: s/one/area/

File: architecture/core.rst [hunk 6]
```diff
@@ -800,42 +800,52 @@ Shared Memory
...
+and  optional extra memory areas for non-contiguous buffers.
```
Comment (jbech-linaro) [inline architecture/core.rst]: s/and  optional/and optional/ (double spaces) s/non-contiguous/noncontiguous/

File: architecture/core.rst [hunk 7]
```diff
@@ -800,42 +800,52 @@ Shared Memory
...
+
+The shared memory area accessed with non-secure rights by OP-TEE core
```
Comment (jbech-linaro) [inline architecture/core.rst]: I'm loosing the context a bit when reading this context.

File: architecture/core.rst [hunk 8]
```diff
@@ -800,42 +800,52 @@ Shared Memory
...
     - Size of the pool
     - Whether or not the memory is cached
 
-The shared memory pool configuration is platform specific. The memory mapping,
-including the area ``MEM_AREA_NSEC_SHM`` (shared memory with non-secure world),
-is retrieved by calling the platform-specific function ``bootcfg_get_memory()``.
-Please refer to this function and the area type ``MEM_AREA_NSEC_SHM`` to see the
-configuration for the platform of interest. The Linux driver will then
-initialize the shared memory pool accordingly.
+Non-contiguous shared buffers
```
Comment (jbech-linaro) [inline architecture/core.rst]: s/Non-contiguous/Noncontiguous/

File: architecture/core.rst [hunk 9]
```diff
@@ -800,42 +800,52 @@ Shared Memory
...
+-----------------------------
 
-.. todo::
+To benefit from non-contigous shared memory buffers, secure world register
```
Comment (jbech-linaro) [inline architecture/core.rst]: s/non-contiguous/noncontiguous/

File: architecture/core.rst [hunk 10]
```diff
@@ -800,42 +800,52 @@ Shared Memory
...
+dynamic shared memory areas and non-secure world must register non-contiguous
```
Comment (jbech-linaro) [inline architecture/core.rst]: s/non-contiguous/noncontiguous/

File: architecture/core.rst [hunk 11]
```diff
@@ -800,42 +800,52 @@ Shared Memory
...
+buffers prior referring to in the OP-TEE API.
```
Comment (jbech-linaro) [inline architecture/core.rst]: s/prior referring to in the OP-TEE API./prior to referring to them using the OP-TEE API/

File: architecture/core.rst [hunk 12]
```diff
@@ -800,42 +800,52 @@ Shared Memory
...
+
+OP-TEE core generic boot sequence discovers dynamic shared areas from the
```
Comment (jbech-linaro) [inline architecture/core.rst]: s/the//

File: architecture/core.rst [hunk 13]
```diff
@@ -800,42 +800,52 @@ Shared Memory
...
+device tree and/or areas explicitly registered by the platform.
+
+Non-secure needs to register buffers as 4kByte chunks lists into OP-TEE core
```
Comment (jbech-linaro) [inline architecture/core.rst]: s/Non-secure/Non-secure side/

File: architecture/core.rst (hunk 1 above)
Comment (etienne-lms) [reply architecture/core.rst]: thanks

File: architecture/core.rst (hunk 2 above)
Comment (etienne-lms) [reply architecture/core.rst]: you expect a change in current proposal?
Maybe change to "Cache attributes of the shared memory buffer"?

File: architecture/core.rst (hunk 3 above)
Comment (etienne-lms) [reply architecture/core.rst]: ok

File: architecture/core.rst (hunk 4 above)
Comment (etienne-lms) [reply architecture/core.rst]: thanks

File: architecture/core.rst (hunk 5 above)
Comment (etienne-lms) [reply architecture/core.rst]: ok

File: architecture/core.rst (hunk 6 above)
Comment (etienne-lms) [reply architecture/core.rst]: thanks + ok for all `noncontiguous`

File: architecture/core.rst (hunk 7 above)
Comment (etienne-lms) [reply architecture/core.rst]: I will remove this sentence. Not really valuable.

File: architecture/core.rst (hunk 12 above)
Comment (etienne-lms) [reply architecture/core.rst]: ok

File: architecture/core.rst (hunk 11 above)
Comment (etienne-lms) [reply architecture/core.rst]: thanks for the proposal.

File: architecture/core.rst (hunk 13 above)
Comment (etienne-lms) [reply architecture/core.rst]: ok

**Comment by jbech-linaro** [comment]
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

**Comment by etienne-lms** [comment]
tag applied

**Comment by jbech-linaro** [comment]
It introduces a couple of warnings, see:
https://travis-ci.org/OP-TEE/optee_docs/jobs/504589879#L558-L566

Not sure why line-numbers are pointing to wrong places, but I can see the same warnings, **but** with correct lines numbers when compiling the docs locally. You can try it out on your own: https://optee.readthedocs.io/building/gits/optee_docs.html#build-optee-docs

**Comment by jbech-linaro** [comment]
And regarding levels, the "order" is like this:
https://optee.readthedocs.io/building/gits/optee_docs.html#sections-chapters

**Comment by etienne-lms** [comment]
Removed review tag. Issue still open, @jbech-linaro comments to address...

**Comment by etienne-lms** [comment]
2 fixup commits for the `make html` issues.

**Comment by etienne-lms** [comment]
This P-R should also fix the following description in [core.rst#using-shared-memory](https://github.com/OP-TEE/optee_docs/blob/0ab5cc2fa933162e0043db7666848ca7fca6b68b/architecture/core.rst#using-shared-memory):
> (...) In such a case, the provided memory must be physically contiguous, since OP-TEE core, who does not handle scatter-gather memory, is able to use the provided range of memory addresses. (...)

**Comment by jbech-linaro** [comment]
@etienne-lms , so this is ready for merge now I suppose. I.e., you don't intend to make any further changes? If that is the case, feel free to add my R-B and squash/rebase the patches.

**Comment by etienne-lms** [comment]
I saw this morning that there is still a description to fix ([above](https://github.com/OP-TEE/optee_docs/pull/5#issuecomment-480727067)). I'll do it straight with the following proposal and apply you tag if you're OK:

"The client application can ask for shared memory allocation using the GlobalPlatform Client API function TEEC_AllocateSharedMemory(...). The client application can also ~~provide shared memory~~ **register a memory buffer** through the GlobalPlatform Client API function TEEC_RegisterSharedMemory(...). ~~In such a case, the provided memory must be physically contiguous, since OP-TEE core, who does not handle scatter-gather memory, is able to use the provided range of memory addresses. Note that the reference count of a shared memory chunk is incremented when shared memory is registered, and initialized to 1 on allocation.~~ **The shared memory reference returned can be used as parameter when invoking a trusted application.**

**Pull request merged**